### ISSUE TITLE
systemd-journal-remote: use distinct directory so persistent journal is not enabled as side-effect

### ISSUE THREAD

**Issue by mbiebl** [systemd/systemd#1035]
The default systemd-journal-remote.service logs to /var/log/journal/remote
This has two effects: Creating that directory, enables persistent logging as a side effect if /var/log/journal did not exist before. Is that really intentional?
I wonder if we shouldn't use two completely distinct directories, like
/var/log/journal and /var/log/journal-remote.

Second, shouldn't systemd-journal-remote.service have a ConditionPathExists=/var/log/journal/remote?

**Comment by mbiebl** [comment]
@keszybz I think this is mostly for you

**Comment by poettering** [comment]
More likely we should just remove Storage=auto support in journald.conf, since it cannot work, due to /var being missing in early boot, and thus resulting in incorrect checks. If people want to turn off persistant storage they should do so via journald.conf, and not via some directories.

Also, /var/log/journal/ is also the place where we put container logs if they are centralized, thus there are many more reasont he dir might exist.

And no, I am very sure that ConditionPathExists= should not be added. This would mean starting the service is a NOP without error message when the dir is missing. And that's just wrong and confusing for a service that is opt-in anyway...

**Comment by mbiebl** [comment]
Fair enough, regarding the Condition, so I dropped that from the bug title.

Still, it seems odd that using remote journal enables the usage of a persistent local journal, as a side-effect, by creating /var/log/journal/remote.

Using distinct directories for remote and local journals (like /var/log/journal and /var/log/journal-remote) would have the benefit what we could keep Storage=auto as default.

**Comment by poettering** [comment]
I am pretty sure we should use a common place for all kinds of journals, and not split that up, in particular since we should provide a way to interleave them. Again, I am pretty sure we should rather remove "Storage=auto" support...

**Comment by mbiebl** [comment]
After dealing with https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=800947 I came to the same conclusion that Storage=auto is actually not that useful and counter-intuitive.
The existing tmpfiles snippet for /var/log/journal is basically useless. Since systemd-journald starts before systemd-tmpfiles on boot, the subdirectory in /var/log/journal will already be created before tmpfiles has a chance to setup an ACL which newly created directories could inherit.

I therefor think we will do the following in Debian [1]:

a/ Ship /var/log/journal in the systemd package
b/ Apply the ACL to /var/log/journal (not the subdirectory) in postinst
c/ Set the default from auto to volatile
d/ If a user had already created a /var/log/journal directory, check for
that in preinst and create a journald.conf.d snippet setting
Storage=persistent
e/ Update the instructions in README.Debian how to enable persistent
journal. Recommend to use a drop-in config in
/etc/systemd/journald.conf.d/ containing

```
[Journal]
Storage=persistent
```

This would also mean, that we can then safely ship /var/log/journal/remote in the package without problems.

The upstream default in d/ should probably be persistent.

[1] https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=800947#65

**Comment by mbiebl** [comment]
The existing tmpfiles snippet for /var/log/journal is basically useless....
... since commit 8b258a645ae63dff3ab8dde6520d2e770e2a40f1

**Comment by poettering** [comment]
@mbiebl so, actually the "auto" mode is not as borked as I claimed.... journald will only write to /var/log/journal _after_ it got one flush request. A flush request is done via SIGUSR1, and via journalctl --flush and the systemd-flush-journal.service service. The flush state is reflected by the existance of /run/systemd/journal/flushed.

Basically, until that file is created journald won't touch /var/log/journal at all, regardess of the mode you pick, even if Mode=persistent. And SIGUSR1 causes it to create that file.

That said, I am not convinced that the auto mode is really such a good idea.

Also, if I'd maintain the stuff in Debian I'd just turn on the persistent mode... I mean, what's the problem, pretty much all daemons keep some stuff in /var, so why would you turn this off for journald?

Also, how come you enable the remote logging stuff anyway by default? People should not install that unless they _actually_ want it, in which case they want persistent logging too.

We really want all stuff under /var/log/journal btw, so that "journalctl -M" can interleave things properly.

**Comment by poettering** [comment]
There's indeed kind of a big contradiction in the whole thing: if you use Mode=persistent but the dir doesn't exist yet, we want journald to create the dir but then again we expect tmpfiles to put the ACL on it (journald cannot do that on its own, since applying ACLs means resolving groups, and NSS is nothing we can do it journald). Now, we'd need tmpfiels to apply the ACL after we created the dir, but before we put the first file in it, so that the default ACLs is used... but that of course cannot work.

Yuck. Not sure how to fix that nicely...

So yeah, mode=persistent is borked if the dir does not exist already... On Fedora, that's what we do, hence we never notice this...

**Comment by mbiebl** [comment]
2015-10-28 1:35 GMT+01:00 Lennart Poettering notifications@github.com:

> So yeah, mode=persistent is borked if the dir does not exist already... On
> Fedora, that's what we do, hence we never notice this...

Right, that's why I suggest  that distros ship /var/log/journal in the
package, set up the ACL on /var/log/journal (not the subdirectory) in the
package postinst phase and we no longer make "auto" the default or remove
it completely.

And then it's pretty easy for distros/users to enable persistent/volatile
logging.
If the /var/log/journal directory is always there and properly setup
without having any side-effects, it's the cleanest solution I think.

## 

Why is it that all of the instruments seeking intelligent life in the
universe are pointed away from Earth?

**Comment by mbiebl** [comment]
2015-10-28 1:31 GMT+01:00 Lennart Poettering notifications@github.com:

> @mbiebl https://github.com/mbiebl so, actually the "auto" mode is not
> as borked as I claimed.... journald will only write to /var/log/journal
> _after_ it got one flush request. A flush request is done via SIGUSR1,
> and via journalctl --flush and the systemd-flush-journal.service service.
> The flush state is reflected by the existance of
> /run/systemd/journal/flushed.

The problem is, that the systemd-journal-flush.service has a
Before=systemd-tmpfiles-setup.service ordering.
So there is a ordering problem/race during boot:
a/ One would have to mkdir /var/log/journal/$machine-id
b/ Then tmpfiles needs to run to apply the proper ACL on that dir
c/ Then systemd-journald-flush needs to run and trigger the creation of the
actual journal files.

This sequence of events is not happening, so the ACL of the *.journal files
will be incorrect. At least as long as they are not rotated aways.

> Also, if I'd maintain the stuff in Debian I'd just turn on the persistent
> mode... I mean, what's the problem, pretty much all daemons keep some stuff
> in /var, so why would you turn this off for journald?

Atm, we still install a syslogger by default (rsyslog) and we want to avoid
having two copies of the log files.
There are still issues with the journal from time to time (files getting
corrupt if the system crashes, journalctl getting confused etc). Basically,
for jessie journald/journalctl didn't seem mature to be the sole provider
for the system logs.

We will revisit that for stretch.

> Also, how come you enable the remote logging stuff anyway by default?
> People should not install that unless they _actually_ want it, in which
> case they want persistent logging too.

We don't enable remote logging my default.  We have a separate binary
package called systemd-journal-remote. And we wanted to ship the
/var/log/journal/remote directory in that package, so the systemd-journal-*
tools would work of the box.

But creating /var/log/journal/{remote} had the undesired side-effect which
has been discussed in this bug report.

> We really want all stuff under /var/log/journal btw, so that "journalctl
> -M" can interleave things properly.

Ok, understood. I'm not arguing about that anymore. Maybe we should retitle
the bug report accordingly then.

## 

Why is it that all of the instruments seeking intelligent life in the
universe are pointed away from Earth?

**Comment by jlebon** [comment]
Apologies for resurrecting an old ticket, though this discussion captures the issue well.

> That said, I am not convinced that the auto mode is really such a good idea.

I think I agree with this.

Let me confirm my understanding here: `systemd-journal-flush.service` today runs before `systemd-tmpfiles-setup.service` so that the former can create the dirs that the latter then needs to add ACLs on? And we *don't* want to create the dirs unconditionally because of the special `Storage=auto` behaviour?

If we drop `Storage=auto`, then there's no reason we couldn't run `systemd-tmpfiles-setup.service` *before* `systemd-journal-flush.service` and thus have the ACLs set up correctly from the get-go, right?

Another example where running before `systemd-tmpfiles-setup.service` causes issues is as in #11903. I.e. right now in Fedora CoreOS, we're relying on an `Z /var - - -` entry to relabel `/var`, but then because the journal runs before the relabeling happens, it triggers SELinux violations while trying to flush.

Hmm, what's the best way to re-order this in FCOS meanwhile? As per `systemd.unit(5)`, drop-ins don't allow resetting the dependency list. So we'd have to either directly override the whole unit, or do a hacky `s/Before=systemd-tmpfiles-setup.service/After=systemd-tmpfiles-setup.service/` so we don't lose other updates to the file.
